src/test/scala/mesosphere/marathon/integration/setup/ProcessOutputToLogStream.scala

```diff
@@ -1,13 +1,58 @@
 package mesosphere.marathon
 package integration.setup
 
-import org.slf4j.LoggerFactory
+import java.time.Clock
+import org.slf4j.{ LoggerFactory, Logger }
+import ProcessOutputToLogStream._
 
 import scala.sys.process.ProcessLogger
+import scala.concurrent.duration._
 
-case class ProcessOutputToLogStream(process: String) extends ProcessLogger {
-  val log = LoggerFactory.getLogger(s"mesosphere.marathon.integration.process.$process")
-  override def out(msg: => String): Unit = log.debug(msg)
-  override def err(msg: => String): Unit = log.warn(msg)
+object ProcessOutputToLogStream {
+  /**
+    * Handles rate limiting logic; calling the method `incrementAndCheck` both registers a log line and returns whether
+    * or not the quota has been exceeded.
+    *
+    * Simple periodic quota mechanism. We keep a tally of the number of log lines for each period. When we enter the
+    * next period, we reset. If we exceed the quota for the period, we start returning false.
+    *
+    * @return Whether or not the quota for this period has been exceeded
+    */
+  final class RateLimiter(log: Logger, maxLinesPerPeriod: Int, period: FiniteDuration, clock: Clock) {
+    private val periodMillis = period.toMillis
+    private[this] var nextPeriodStart = System.currentTimeMillis() + periodMillis
+    private[this] var linesThisPeriod = 0
+    def incrementAndCheck(): Boolean = synchronized {
+      val now = clock.millis()
+      if (now >= nextPeriodStart) {
+        val suppressedLines = linesThisPeriod - maxLinesPerPeriod
+        if (suppressedLines > 0)
+          log.error(s"Logger was rate-limited; ${suppressedLines} lines were supressed")
+
+        linesThisPeriod = 1
+        nextPeriodStart = now + periodMillis
+        false
+      } else {
+        linesThisPeriod += 1
+        linesThisPeriod > maxLinesPerPeriod
+      }
+    }
+  }
+}
+
+final case class ProcessOutputToLogStream(process: String, maxLinesPerPeriod: Int = 100, period: FiniteDuration = 1.second) extends ProcessLogger {
+  private val log: Logger = LoggerFactory.getLogger(s"mesosphere.marathon.integration.process.$process")
+  private val rateLimiter = new RateLimiter(log, maxLinesPerPeriod, period, Clock.systemUTC())
+
+  override def out(getMsg: => String): Unit = if (log.isDebugEnabled()) {
```

Comment:
I might be wrong but shouldn't this line be thread safe as well? I do not see an apparent issue but just checking.

Reply:
Looks pretty thread-safe to me.

Reply:
Yeah, it's fine; rateLimiter is thread safe. `out` and `err` are independently called serially, but, as I understand it, could each be called concurrently.